capability.go

```diff
@@ -136,6 +139,7 @@ func NewClient(hook ClientHook) *Client {
 		refs:       1,
 		resolved:   newClosedSignal(),
 	}
+	(&h.metadata).init()
```

Comment:
This more of a design/devex issue than anything else, but I feel that this should be handled by a constructor.  As it is, it adds visual noise to the code-base and raises the possibility of forgetting to call `init()`. 

In fact, I think `clientHook` should probably have a corresponding `newClientHook` function, but we don't have to address that in this PR.

Reply:
The reason I did it this way is I wanted to have the `Metadata` live directly in the `clientHook`. There is `NewMetadata`, but it returns a pointer...

Reply:
That's what I gathered, but couldn't we nevertheless return a concrete struct from the constructor?  It would copy the mutex, but there would no longer be any copies after first use.  Unless I'm misremembering, the Go toolchain should be perfectly happy with that.

Reply:
Ok, changed (I also tweaked the wording on the comment so it's clear it's ok to copy before first use, which addresses my main hangup).

I also merged in main and used syncutil.With in a couple places where it's appropriate.